lib/signer.ex

```diff
@@ -40,78 +40,85 @@ defmodule AntikytheraAws.Signer do
     - `params` must be decoded query strings, in a list of tuple-2
         - To convert from `Antikythera.Http.QueryParams` (which is `Croma.SubtypeOfMap`), just use `Map.to_list/1`
 
-    Note: "content-type" and "x-amz-content-sha256" headers are required for S3, but not generally.
+    Note: "content-type" header is required for S3, but not generally.
     """
     defun prepare_headers(%Creds{access_key_id: aki, secret_access_key: sak, security_token: st} = _creds,
                           region  :: v[String.t],
                           service :: v[String.t],
                           method  :: v[Method.t],
                           path    :: v[UPath.t],
-                          payload :: v[String.t],
+                          payload :: v[binary],
                           headers :: v[Headers.t],
                           params  :: [{String.t, String.t}]) :: Headers.t do
       if String.contains?(path, "?"), do: raise("`path` must not include query string part!")
+      payload_sha256                          = hex_sha256(payload)
+      {sign_ready_headers, amz_date}          = canonicalize_headers_with_amz_date(headers, st, service, payload_sha256)
+      {canonical_request, signed_headers_str} = make_canonical_request_with_shs(sign_ready_headers, method, path, payload_sha256, params)
+      signature                               = compute_signature(aki, sak, amz_date, region, service, canonical_request, signed_headers_str)
+
+      Map.put(sign_ready_headers, "authorization", signature)
+    end
+
+    defunp canonicalize_headers_with_amz_date(headers        :: v[Headers.t],
+                                              st             :: v[nil | String.t],
+                                              service        :: v[String.t],
+                                              payload_sha256 :: v[String.t]) :: {Headers.t, AmzDate.t} do
       downcased_headers = Map.new(headers, fn {key, val} -> {String.downcase(key), val} end)
       if !Map.has_key?(downcased_headers, "host"), do: raise("'host' header is required!")
-      {amz_date, headers_with_date} = case downcased_headers do
-        %{"x-amz-date" => ad} -> {ad, downcased_headers}
+      {amz_date, headers_with_date} = gen_amz_date_and_put_date(downcased_headers)
+      headers_might_with_st         = if st, do: Map.put(headers_with_date, "x-amz-security-token", st), else: headers_with_date
```

Comment:
This usage of "might" isn't grammatically correct (where's the verb?).
Change the variable name to e.g. `headers_possibly_with_st`.
Thanks to the refactoring effort its scope is fairly limited, so IMO it's also possible to use stupid names such as `headers1`, `headers2`, ... here.